Auto-assign RF type based on feature data type

grovduck opened the pull request:
This PR is the first step toward closing #87 and provides support for assigning the type of scikit-learn random forest (either `RandomForestRegressor` or `RandomForestClassifier`) based on each feature's datatype.  At this point, here is the general mapping of data type to random forest type:
- floating-point -> "regression"
- integer -> "regression"
- pd.Category -> "classification"
- string -> "classification"
- object -> "classification"

Once features are assigned to random forest types, a separate forest is built for each feature.  In a future PR, we will support the option for users to override the random forest type associated with specific features.

grovduck commented:
It's embarrassing when Copilot writes a better description of the PR than I do ...

aazuspan commented:
@grovduck, I think the overall design looks solid! Awesome start to a complex feature. I do have some thoughts on your points above and have a long response drafted, but I realized I might be a little confused and just wanted to be sure I'm on the right page before diving deeper into the details.

I've been running some quick tests just to confirm that I understand how this works, and haven't been able to successfully fit a model that uses an RF classifier due to some errors that show up with weighting in the `KNeighborsRegressor`. For example:

```python
from sknnr import RFNNRegressor
from sknnr.datasets import load_moscow_stjoes

X, y= load_moscow_stjoes(return_X_y=True, as_frame=True)

# Create some dummy columns that should be fit with classifiers
y["str_col"] = y.idxmax(axis=1)
y["cat_col"] = pd.cut(y.PSME_BA, 3, labels=["low", "mod", "high"])
RFNNRegressor(weights="uniform").fit(X, y)
```

Using `uniform` and `distance` weighting return slightly different errors, but both seem to arise from the weighting functions expecting numerical values in the `self._y` attribute, from what I can tell at a quick glance.

Is that something you've run into, or am I doing something wrong?

grovduck commented:
@aazuspan, that's a great point and definitely needs more code and tests.  I was doing something very similar when developing, but I was creating an ancillary `y_fit` to store the non-numeric attributes.  
```
from sknnr import RFNNRegressor
from sknnr.datasets import load_moscow_stjoes

X, y= load_moscow_stjoes(return_X_y=True, as_frame=True)
y_fit = y.copy()

# Create some dummy columns that should be fit with classifiers
y_fit["str_col"] = y_fit.idxmax(axis=1)
y_fit["cat_col"] = pd.cut(y_fit.PSME_BA, 3, labels=["low", "mod", "high"])
est = RFNNRegressor(weights="uniform").fit(X, y, y_fit=y_fit)
print(est.predict(X))
```
For the record, I think we'd have similar issues when using non-numeric columns in other estimators right now (e.g. `GNNRegressor`), but they would bonk in the transformer rather than in neighbor-finding.  It's just that these non-numeric features are explicitly allowed for `RFNodeTransformer`.  I think that's why this estimator is such a bear!
```
from sknnr import GNNRegressor
from sknnr.datasets import load_moscow_stjoes

X, y= load_moscow_stjoes(return_X_y=True, as_frame=True)

# Create some dummy columns that should be fit with classifiers
y["str_col"] = y.idxmax(axis=1)
y["cat_col"] = pd.cut(y.PSME_BA, 3, labels=["low", "mod", "high"])

# This raises TypeError in _cca.py
GNNRegressor(weights="uniform").fit(X, y)
```
Thanks so much for digging in and happy to modify however you see fit.

aazuspan commented:
Ah, that makes perfect sense. I hadn't fully grokked that the classification targets are only used for neighbor-finding via `y_fit`, but of course that's the case since we can't apply the underlying regression estimator to categorical data. Thanks for the explanation!

Now that I'm finally on the same page, here are some thoughts on your original design points. 

> I don't think I have an exhaustive list of all the different datatypes we may encounter. In rfnode_transformer.py, I've created the DTYPE_TO_RF_TYPE dictionary to start this list. I have most of the common numpy types we'd expect to see (although certainly not exhaustive) as well as some of the special pandas extension dtypes. Anything that isn't explicitly in this list is unsupported, so we may be excluding some exotic types that would be fine in random forests. I've hard coded this list for explicitness, but perhaps there is a more elegant way to handle this.

If you feel strongly about sticking with strict explicitly supported types, I think the mapping makes sense. We could simplify a little bit by checking lowercase keys so we don't need to worry about case-sensitivity, and should probably also include rarer types like bools and 128 bit ints and floats.

If we want to go a little more implicit but equally strict, we could replace the numeric type mappings with an `np.issubdtype(dtype, np.number)` check, which should pass for any integer, floating point, or complex number. That would require reworking things a little as it would need the dtype instead of its name, and we'd need to catch errors for unrecognized dtypes like categorical.

If you're not tied to the strict supported types, I'd be tempted to just throw any non-numeric types into a classifier and let `sklearn` decide if it's valid, for the sake of simplicity and flexibility, although your list probably already covers 99% of cases.

> In order to avoid bringing in pandas as a dependency, I have two awkward private methods in RFNodeTransformer that duck-type pd.DataFrame and pd.Series and that feels like the wrong way to go about it.

The optional `pandas` dependency does complicate some things - I'm willing to bend on that if it would save us some effort. I imagine almost everyone using `sklearn` is already using `pandas` anyways.

With that said, I still think the duck typing approach was a good choice, although perhaps we could be more lenient with the check by just looking for the necessary attrs like `columns` and `dtypes` rather than explicitly requiring Pandas types? That would make it easier to potentially support other dataframe packages in the future if they follow a compatible API.

> As of now, the method _get_column_names gets column names for the pandas types and creates strings of column index for numpy arrays. Then in _set_rf_types, I use the ordered lists of column names and dtypes and go through some gyrations to ultimately create a column index (int) to rf type (str) dictionary. 

Mapping from indexes to RF types as the final step makes sense, but if it was possible to store `rf_type_dict_` before doing that conversion from column names to indexes, that would be helpful to allow checking RF types after fitting with a dataframe, without needing to crosswalk indexes back to names. Probably not worth the effort if it would be a big lift, though.

> I think we need to keep it this way in order to accommodate the possible user overrides (currently as a TODO in the _set_rf_types function.

Good forward thinking to set up the groundwork for user overrides!

> So now I'm creating a y_arr object out of that _validate_data call, because it has to be called before the auto-mapping. Yuck!

I was a little confused when I first saw this, but the justification makes sense! I guess we could flip it and store `y_df` while converting `y` to array, but I'm not sure that's any clearer. I think your solution looks good here!

I haven't put any thought into the testing yet, so I'll hold off on commenting on that for now. 

The only other thing that popped out to me so far is that some of the `RFNodeTransformer` methods like the Pandas checks and column processing could perhaps be static/class methods or utility functions?

grovduck commented:
Thanks for the great feedback, @aazuspan.  Will tackle this next week.

grovduck commented:
I realized that I was typically using the term "feature" when referring to `y`-based columns.  I decided that using "target" was a bit more accurate.

aazuspan commented on src/sknnr/utils/__init__.py:
```diff
@@ -0,0 +1,70 @@
+        return []
+    if obj.ndim == 1:
+        obj = obj.reshape(-1, 1)
+    return [type(obj[0, k]) for k in range(obj.shape[1])]
```
Just thinking out loud, it seems like there should be a way to have Numpy infer the dtype of each column to avoid the potential for mismatched types within a column to return inaccurate results. Nothing immediately pops out to me though, and I don't think that's really a big concern. Maybe you could create a separate array from each column and check the inferred dtype, but accessing columns without creating the 2D array first seems tricky...

Feel free to ignore this comment, since I don't have any clear ideas and I think it's fine as-is.

grovduck replied:
No, I think this is an important point and one that I took the easy way out.  Definitely related to the above comment, but now suppose our list is:
```
>>> lst = [[1, "a", 0.3], ["b", 0.4, 2]]
```
This should optimally raise an error or at least warn on mismatched data types.  But right now, only checking the top row is not catching this (although I'm guessing the estimators will either bonk or upcast each column).  

One option is that we utilize numpy record arrays to help with typing.
```
>>> good_lst = [(1, "a", 0.3), (2, "b", 0.4)]
>>> bad_lst = [(1, "a", 0.3), ("b", 0.4, 2)]
>>> def lst_to_recarray(lst):
...     # Get the assumed dtype from the first row
...     dtype = np.rec.fromrecords(lst[0]).dtype
...     # Convert the full LoL using this dtype
...     # Raise on type mismatches
...     try:
...         return np.rec.fromrecords(lst, dtype=dtype)
...     except ValueError:
...         raise ValueError("One column had mismatched data types")
...
>>> lst_to_recarray(good_lst)
rec.array([(1, 'a', 0.3), (2, 'b', 0.4)],
          dtype=[('f0', '<i8'), ('f1', '<U1'), ('f2', '<f8')])
>>> lst_to_recarray(bad_lst)
<snip>
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 9, in lst_to_recarray
ValueError: One column had mismatched data types
```
This approach has a couple complications:
- Passing list of lists is being deprecated - one should pass lists of tuples or tuple of tuples, so we'd have to manage that before passing
- We'd need to convert the output to a regular 2D numpy array, but I think this is trivial using `np.asarray(recarray, dtype=object)`

If you have other thoughts, I'd totally welcome them.  Dtypes are hard!

aazuspan replied:
Record arrays are new to me, but look very interesting. That does seem like a good option to enforce consistency in the dtypes, although I wasn't quite clear if the dtypes can be retrieved from the record array or if that would be a validation step followed by the existing approach of extracting the type of the first elements (now that we know they match the column)? 

What I vaguely had in mind was less thorough and more permissive by just building a new array from each column and letting Numpy promote and infer the dtype. Whether that's a better approach is definitely debatable, but something like below seems to work (I only modified the last line of your current function):

```python
def get_feature_dtypes(obj) -> list[np.dtype]:
    ...

    obj = np.asarray(obj, dtype=object)
    if len(obj) == 0:
        return []
    if obj.ndim == 1:
        obj = obj.reshape(-1, 1)

    return [np.array(obj[:, i].tolist()).dtype for i in range(obj.shape[1])]

>>> get_feature_dtypes(good_lst)
[dtype('int64'), dtype('<U1'), dtype('float64')]
>>> get_feature_dtypes(bad_lst)
[dtype('<U21'), dtype('<U32'), dtype('float64')]
```

I'm happy however you choose to tackle this, including the current approach, so don't let me derail you!

grovduck replied:
NOTE: We eventually with @aazuspan's approach of promoting each target, but with checks to ensure that string-types and non-string-types did not co-occur in a target.  See discussion below.

aazuspan reviewed (changes requested):
The changes look great, @grovduck! Just a few more comments from me, but it looks like it's getting very close (aside from tests which I still haven't focused on).

> np.issubdtype works for many dtypes, but not for pandas extension types. For that, I rely on the kind attribute. I imagine that there could still be types that would raise on both these tests, but I didn't yet come across any in the test suite.

Good find! I don't think it's worth worrying too much about really rare edge cases. If they come up, we can always address them then.

> I've kept bool as "classification" as I think that is likely the most common use case.

Good call, I agree!

> I could definitely use some help in typing the parameter t in this function.

This is tricky. Since those values can originate from the `dtypes` of any dataframe-like, I don't think we can confidently constrain that to a specific set of types. Most cases would probably be covered by `typing.Type | np.dtype | pd.core.dtypes.dtypes.ExtensionDtype`, but I think it's more accurate just to stick with `Any`, unfortunately.

> I chose to keep this method in RFNodeTransformer as I didn't know if it had broader utility, but if you think it should be external, that's good with me.

I think it's general-purpose enough that it could be a utility, but I'm fine wherever you'd rather keep it!

> Ultimately, I was able to avoid creating a y_arr copy with a little reorganization. The function get_feature_names_and_dtypes gets everything it needs to allow user overrides (I hope!) before converting y to an array. Much cleaner.

Nicely done! I didn't think of that, but it's a much better solution.

grovduck commented on src/sknnr/utils/__init__.py:
```diff
@@ -0,0 +1,70 @@
+    if is_series_like(obj):
+        return ["0"] if obj.name is None else [obj.name]
+    obj = np.asarray(obj, dtype=object)
+    if len(obj) == 0:
+        return []
```
It sure seems like that, doesn't it!  I wrestled with this one for a while.  If you take out this test, it fails one of the transformer checks:
```
FAILED tests/test_transformers.py::test_sklearn_transformer_checks[RFNodeTransformer()-check_estimators_empty_data_messages] - AssertionError: The estimator RFNodeTransformer does not raise a ValueError when an empty data is used to train. Perhaps use check_array in train.
```
I think what happens in this test is that the transformer is passed `[]` as `y`.  In this function, I'm converting all sequences (lists and tuples) into numpy arrays so that I can use both `reshape` and `shape` later on.  But then, because `np.array([]).ndim == 1`, it reshapes it to (0, 1).  Then the list comprehension at the end can't index [0, 0] and raises an `IndexError`.  So I think the error message is saying that it doesn't receive the `ValueError` because it's getting the `IndexError` first.

Clumsily, I have that `len(obj) == 0` to  catch this, but if there is a better way, I'm happy to rewrite.

aazuspan replied:
Ah, the `IndexError` with empty data makes sense, thanks for explaining! I think the `len` check is a good solution.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,36 +150,87 @@ def __init__(
+                msg = f"Unsupported type {t}"
+                raise TypeError(msg) from err
+
+    def _set_rf_types(self, target_info: dict[str, Any]) -> dict[str, str]:
```
That's even better!

grovduck commented on src/sknnr/utils/__init__.py:
```diff
@@ -0,0 +1,70 @@
+        return list(obj.columns)
+    if is_series_like(obj):
+        return ["0"] if obj.name is None else [obj.name]
+    obj = np.asarray(obj, dtype=object)
```
Another one I struggled with!  Check my logic here.  If we have a list like:
```
>>> lst = [[1, "a", 0.3], [2, "b", 0.4]]
```
If we don't include `dtype=object`, all columns will get upcast:
```
>>> np.asarray(lst)
array([['1', 'a', '0.3'],
       ['2', 'b', '0.4']], dtype='<U32')
```
but, using `dtype=object` correctly sets types on each column:
```
>>> arr = np.asarray(lst, dtype="object")
>>> arr
array([[1, 'a', 0.3],
       [2, 'b', 0.4]], dtype=object)
>>> list(type(arr[0, i]) for i in range(arr.shape[1]))
[<class 'int'>, <class 'str'>, <class 'float'>]
```
There might be an easier way to do this, though, that I'm unaware of.

aazuspan replied:
Ah, you're totally right, of course. I misunderstood how the `object` dtype works, but now that I bothered to read [the docs](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.object_) this makes sense, and is probably an efficient solution since the array just stores references to the underlying objects.

Thanks for the clear example!

grovduck commented:
@aazuspan, thanks for another helpful look!  Still a number of unresolved issues, but getting better.

> > I chose to keep this method in RFNodeTransformer as I didn't know if it had broader utility, but if you think it should be external, that's good with me.
> 
> I think it's general-purpose enough that it could be a utility, but I'm fine wherever you'd rather keep it!

I've moved this into `utils` as well.

grovduck commented:
Hey @aazuspan, another big dump of changes, but at least I've written a few tests to accommodate these changes!

I think the main outstanding point was how to characterize and handle features (really targets) that had mixed datatypes.  This is probably pretty uncommon in the first place, so likely this is going way overboard.  I ended going with your suggestion:

> What I vaguely had in mind was less thorough and more permissive by just building a new array from each column and letting Numpy promote and infer the dtype.

I still wanted the user to get a warning when mixed types were found in a feature, so currently that's raising a `UserWarning`.  But the dtype assigned to a feature is the type that is "smallest type that can represent all elements in that feature".  

The ugliest part of the code right now is that `get_feature_dtypes` is really only responsible for setting the dtypes, but if there are mixed types in a feature, we don't upcast to that type in this function.  So, I've added a return value called `upcast_required` that signals whether or not the `y` data needs to be upcast before fitting the individual forests.

Take the case of a single string in an integer feature/target.  The dtype gets captured as "<U*" and the random forest type is subsequently set to "classification".  However, at this point, the `y` data hadn't yet been converted so it's still a mix of the ints and the one str.  When building a `RandomForestClassifier`, sklearn uses the function [`type_of_target`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.multiclass.type_of_target.html) to apply a heuristic on what type of data it thinks it has.  In a case like this, it outputs "unknown" and the forest can't be built.  

So now, before building the forests, I convert each target in `y` to its upcast datatype if needed.  I can't just do it for *all* `y` arrays because there may be `pandas` datatypes stored in `target_info` and, at this point, `y` is a numpy array.

I'm sorry this is turning into such a beast.  If you think that we shouldn't put these types of guardrails up, I totally understand and I'm guessing it will simplify the code quite a bit.  I think we would just want to error early on any `y` that had a target with mixed data types.

aazuspan reviewed (commented):
@grovduck, the tests look awesome. I like the granular utility tests combined with the broader transformer tests to ensure they're working as expected in practice. 

The only other thing that I can think to test is that RFNN works end-to-end with one or more classification forests. Maybe it would it be worth adding a new regression test for RFNN with a realistic case where some of the features are binned into categoricals? Including some regression/classification-specific RF parameters in that could also add coverage for the parameter specialization. 

> The ugliest part of the code right now is that get_feature_dtypes is really only responsible for setting the dtypes, but if there are mixed types in a feature, we don't upcast to that type in this function. So, I've added a return value called upcast_required that signals whether or not the y data needs to be upcast before fitting the individual forests.

I hadn't considered the complexity of promoting dtypes in `get_feature_dtypes` when I suggested that - sorry to throw such a big wrench into this! I wasn't familiar with the `type_of_targets` check, but it makes sense that `sklearn` rejects those mixed types. 

I do think that I would lean towards a more hands-off approach with mixed types and probably avoid the upcasting, but feel free to overrule that. My thinking is that mixed types would almost always be a mistake, and would raise an error if you tried to use them with any other classifier (I think - correct me if I'm wrong). The unique thing here is that users may not realize that they're using a classifier under-the-hood, so I think you're right that we should error early and make the error clearer than the default "Unknown label type" raised by `sklearn`. 

If we did decide to go with an early-error approach, would it be practical for us to simply try to fit the classifiers, catch any mixed-type errors, and re-raise with a better message? I haven't thought that through so it may be problematic, but I'm a little concerned about the performance of the `unique_types` check with large datasets. I haven't done any benchmarks though, so maybe that's negligible compared to actually fitting the forests.

grovduck commented on tests/test_transformers.py:
```diff
@@ -183,3 +185,33 @@ def test_transformers_raise_out_of_range_n_components(transformer, n_components)
+        y = pd.Series(y)
+        y.iloc[-1] = "mixed"
+    else:
+        y[-1] = "mixed"
```
Good catch and much nicer!  I've got this incorporated in the newest version.

grovduck commented:
@aazuspan, sorry that this was so long in coming.  I had a couple of days with false starts on this, but I feel like we've landed *close* to the right spot with your help.

> I hadn't considered the complexity of promoting dtypes in `get_feature_dtypes` when I suggested that - sorry to throw such a big wrench into this! I wasn't familiar with the `type_of_targets` check, but it makes sense that `sklearn` rejects those mixed types.

As you'll see with the recent changes I've made, I think promoting dtypes is still the correct way to go.  For RFNN, I think we need to consider `y` as a list of univariate targets since that is the way we are modeling with them.  Because our targets can have mixed dtypes by *column*, but shouldn't by *row*, I was having issues making `validate_data` do what we wanted it to do.  By calling `validate_data` with `y` that had a dtype of `object`, this correctly allowed the different dtypes by column, but wasn't picking up any issues with mixed types within a column.

What I've done now is to validate `X` separately and introduce a new internal `_set_targets` method in `RFNodeTransformer`.  This method does the following:
1. Gets feature dtypes via `utils.get_feature_names_and_dtypes`.  I have modified `get_feature_dtypes` to return the minimum *numpy* dtype that holds all elements, with the one exception that if a target is of type `pd.Categorical`, it will retain this type.  Because we eventually need to convert all data to numpy arrays, I even overwrite the pandas extension dtypes to be numpy dtypes.
2. Iterate over targets doing the following:
    1. Ensure that the target has no NaN-like data elements.  At this point, this could be `np.nan`, `None`, or even `pd.NA`.  We raise an error if any NaN-like elements are found.
    2. Handle dtype conversions to the target
        1. Ensure that the target is not a mix of numeric and non-numeric data.  I think the only mixed types that we *don't* want to allow are between numeric and non-numeric data, as promotion from ints to floats seems like it should be OK (although I'm OK revisiting this decsion).  If the promoted type from the call to `get_feature_dtypes` is string-like and we have non-strings in the target, we raise an exception.  (Here I'm using `any` rather than `np.any` in my check which I'm now realizing might not be the right choice, so we might want to vectorize that?)
        2. If the dtype is `pd.Categorical`, retain the type information, but promote the data to the minimum numpy dtype
        3. If neither of these apply, just promote the data to the identified (and possibly promoted) numpy dtype
    3. Call `check_array` on the target.  I'm not sure this is necessary other than it ensures consistency with whatever constraints the estimator might have.
3. Return the targets as a list of 1-d numpy arrays as well as the target information.

> I do think that I would lean towards a more hands-off approach with mixed types and probably avoid the upcasting, but feel free to overrule that. My thinking is that mixed types would almost always be a mistake, and would raise an error if you tried to use them with any other classifier (I think - correct me if I'm wrong). The unique thing here is that users may not realize that they're using a classifier under-the-hood, so I think you're right that we should error early and make the error clearer than the default "Unknown label type" raised by `sklearn`.

I probably did a mix of what you suggested here.  We are erroring out early on mixed data types within targets, but I wouldn't call it a hands-off approach!

> If we did decide to go with an early-error approach, would it be practical for us to simply try to fit the classifiers, catch any mixed-type errors, and re-raise with a better message? I haven't thought that through so it may be problematic, but I'm a little concerned about the performance of the `unique_types` check with large datasets. I haven't done any benchmarks though, so maybe that's negligible compared to actually fitting the forests.

This was a great idea and I tried to make this work because it would be much more hands-off, but this was the source of my earliest false start.  Take this simple example:
```
y = np.array([[1.1, 2.3, 3.2], [True, False, True], [1, "mixed", 3]], dtype=object).T
```
We have to use dtype=object to maintain the different target dtypes.  We want the first target to be a regression forest and the second to be a classification forest and the third is obviously a mixed type.  Without any promotion of target dtype, we could try to just create the forests by target.   However, when trying to create a classification forest on the second target, it must call `type_of_target`:
```
>>> type_of_target(y[:, -1])
'unknown'
```
This is because the dtype is `object` rather than `bool`.  Ultimately I decided that promotion to the minimum numpy dtype was necessary, but I still may be mistaken in that thinking.
  
> The only other thing that I can think to test is that RFNN works end-to-end with one or more classification forests. Maybe it would it be worth adding a new regression test for RFNN with a realistic case where some of the features are binned into categoricals? Including some regression/classification-specific RF parameters in that could also add coverage for the parameter specialization.

Another great suggestion.  I've created a simple two-target `y_fit` using the total basal area as a regression target and the species with the max basal area as a classification target and parameterized over a few combinations that we're already using which adds a few regression test files.  I like the idea of also varying over regression/classification-specific RF params (e.g. `criterion` or `max_features`) but worried that the number of combinations I've already introduced in this test would further explode.  Any thoughts on how best to handle this?

aazuspan commented:
@grovduck, apologies as this isn't a real review yet, and I haven't addressed all of your points. I've been trying to wrap my head around the complexity of validation, and I think I may finally get it, but want to confirm before I dive any deeper and potentially push us further off course!

> For RFNN, I think we need to consider y as a list of univariate targets since that is the way we are modeling with them. Because our targets can have mixed dtypes by column, but shouldn't by row, I was having issues making validate_data do what we wanted it to do.

I was struggling to understand why we can't rely on `validate_data` and `type_of_target` to fit classifiers, and I think I may finally get it. I need to write out this logic step-by-step for my own benefit, so apologies if this is what you've been trying to explain to me all along:

1. `validate_data(y=y)` will always return an `object` array if there are any mixed types in `y`.
2. `type_of_target(y)` will always return "unknown" for an `object` array (unless the elements are strings).
3. Fitting an `sklearn` classifier will always raise an error if `type_of_target(y)` returns "unknown".

Therefore, fitting a `RandomForestClassifier` with one column of an array or dataframe with mixed types will fail if that column is not a string. For example:

```python
import numpy as np
from sklearn.ensemble import RandomForestClassifier

X = np.random.rand(4, 3)
y = np.array([[1.1, 2.3, 3.2, 1.1], [True, False, True, True]], dtype=object).T

# This is essentially what happens internally in RFNNRegressor as it builds each forest
# ValueError: Unknown label type: unknown. 
RandomForestClassifier().fit(X, y[:, -1])
```

As a result, we can't fit an `RFNNRegressor` with a boolean `y_fit` column while relying only on `validate_data`. If I understand right, you've solved that by splitting `y` into a list of column arrays and setting the dtype of each column to the type inferred by `numpy`, thereby converting an array of booleans to a dtype of `bool` instead of `object`, allowing `type_of_target` to correctly return "binary" instead of "unknown", and allowing the classifier to be fit. However, you need to catch mixed type errors there to avoid accidentally upcasting numbers to strings.

Sorry it's taken me so long, but am I finally understanding the situation, and why we need `_set_targets` to both modify dtypes *and* catch mixed types?

>  I've created a simple two-target y_fit using the total basal area as a regression target and the species with the max basal area as a classification target and parameterized over a few combinations that we're already using which adds a few regression test files. 

I can see the motivation behind parameterizing `test_mixed_rfnn_forests` for safety, but I wonder if that may be unnecessary since those parameters are already covered for RFNN by `test_predict` and `test_kneighbors`? For example, I can't think of a realistic way that `return_dataframe_index` would break *only* when a classification forest is included, since it's unrelated to the forest type. If you'd rather leave those in though, I'm okay with that. It's a relatively small number of tests in the scheme of things.

> I like the idea of also varying over regression/classification-specific RF params (e.g. criterion or max_features) but worried that the number of combinations I've already introduced in this test would further explode. Any thoughts on how best to handle this?

My thought was just to hard-code a fixed value into the test with the idea that we know they're currently working, and the regression test would catch it if they stop working as it would change the result. The downside is that a regression test alone wouldn't necessarily make it clear *why* they broke, so another alternative would be an explicit test that checks the parameters of the forests.

aazuspan commented:
Thanks for the clarifications! Now that I'm on the same page, I'll go back through and do a full code review (hopefully tomorrow). Thanks for being patient.

> I agree with your point of not knowing why they broke, but I'm afraid I'm not quite understanding what the test would be.

I kind of glossed over my thinking here. The potential bug I'm thinking about is that you could comment all the lines below without failing any tests:

https://github.com/lemma-osu/sknnr/blob/1f94081143f5619a7f0d01ac394483c9dcb0986c/src/sknnr/transformers/_rfnode_transformer.py#L256-L257

https://github.com/lemma-osu/sknnr/blob/1f94081143f5619a7f0d01ac394483c9dcb0986c/src/sknnr/transformers/_rfnode_transformer.py#L261-L263

Those aren't implicitly covered by `test_mixed_rfnn_forests` because it passes default values for the corresponding forest types, so omitting the parameters doesn't change the results. We could add implicit coverage by passing non-default values (e.g. `criterion_reg="absolute_error"`, as that would ensure we get different results if the parameter isn't correctly passed. The downside of implicit coverage is that it wouldn't be clear why the regression test failed if something went wrong.

 If we wanted more explicit coverage, we could do something like:

```python
# Fit with a combination of regression and classification forests
est = RFNodeTransformer(criterion_reg="absolute_error", criterion_clf="gini").fit(X, y)

# Confirm that all the specialized parameters are set on the correct forests
for rf in est.rfs_:
    if isinstance(rf, RandomForestClassifier):
        assert rf.get_params()["criterion"] == "gini"
    else:
        assert rf.get_params()["criterion"] == "absolute_error"
```

There's probably a cleaner way to put that together to handle multiple parameters for each RF type, but that's what I was roughly thinking.

I'm not sure we'd gain anything by parametrizing over `criterion_reg` in the regression test. As long as we're correctly passing it to the forest, I think it's up to `sklearn` to use it.

aazuspan commented on src/sknnr/utils/__init__.py:
```diff
@@ -0,0 +1,137 @@
+    return bool(
+        x is None
+        or (isinstance(x, float) and np.isnan(x))
+        or x.__class__.__name__ == "NAType"
```
Worth a quick note that this covers `pd.NA`?

grovduck replied:
I've expanded the docstring to be more explicit here.

Good question on the `float` check.  Strings definitely don't work with `np.isnan` and I guess I was thinking that we're trying to catch `np.nan` (or anything that evaluates to `np.nan - e.g. `np.log(-1)`) and those would always be float type.  Is my thinking correct on that?

aazuspan replied:
Makes sense, thanks for explaining!

aazuspan reviewed (changes requested):
@grovduck, I think the overall design is looking really solid, especially considering all the sharp edges and complications around arrays and dataframes and mixed types. My suggestions mostly focus on documentation to help me remember how this works in the future.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
+        for i, (name, dtype) in enumerate(target_info.items()):
+            target = y[:, i]
+
+            if np.any(v_is_nan_like(target)):
+                raise ValueError(f"Target {name} has NaN-like elements.")
```
I've added a comment here that hopefully captures the gist of what you were looking for.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
+            # 1. If the dtype is pd.CategoricalDtype, promote the data but keep
+            #    the dtype as category.
+            # 2. If the target is a string dtype, but has non-string elements,
+            #    this represents mixed types that shouldn't be allowed.
```
Ooo boy, I struggled with this one.  I'm not sure my logic is totally solid on this one, but what you've surmised is basically my thinking (`bool` -> `int`, `int` -> `float` are OK, but `int` -> `str`, `float` -> `str` are not).  

I think I was concerned about the downstream effect of how these will be cast into forests, e.g. numeric data will flow to regression, non-numeric (string) data will flow to classification.  But even that isn't terribly consistent because you could potentially have a target like `[True, False, 1.]` that _looks like_ a `bool` but will be treated like a `float`.  I feel like these extreme corner cases are leading us down rabbit-holes.

The other question I had was about your comments about lossy/lossless.  I think I'm missing something, because `int` to `float` can be lossy given a large enough magnitude `int` and `int` to `str` is lossless.  But I think I'm missing your intent on this.  For now, my comments use "lossy/lossless" terminology and incorporate what I _think_ the intent is, but I don't think I've necessarily hit the mark here.

aazuspan replied:
> I think I was concerned about the downstream effect of how these will be cast into forests, e.g. numeric data will flow to regression, non-numeric (string) data will flow to classification.

That seems like a good approach to me, and I agree with your casting logic. 

> int to str is lossless.

Yeah, you've got me there. I was thinking of this as "semantically lossless" (a term I just made up) in the context of how it affects building the forest (like you described), but since there's a well-established definition for lossless that this *doesn't* fit, maybe we're better off using another term? The first thing that comes to mind is "safe promotion" instead of "non-lossy promotion". It's a little vague, but I think the examples you provide make the meaning clear.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
+                not np.issubdtype(type(v), np.str_) for v in target
+            ):
+                raise ValueError(
+                    f"Target {name} has mixed types and cannot be converted to "
```
See my comment below around lossy/lossless and how I'm not quite understanding that.  Great idea on adding information about the mixed dtypes in the error message.  I've changed the terminology a bit to be explicit about string-like and non-string-like rather than using "mixed"

aazuspan replied:
Agreed, lossless wasn't accurate! We could go for something more vague but accurate like "cannot be safely converted"?

grovduck replied:
Great, I've replaced all lossy/lossless terminology with unsafe/safe.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
 
-    def fit(self, X, y):
-        _, y = _validate_data(self, X=X, y=y, reset=True, multi_output=True)
+    def _set_targets(
```
Agreed.  I changed this to `_validate_and_promote_targets`.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
+        _validate_data(self, X=X, reset=True)
+
+        # Set targets and standardize dtypes within `y`
+        y, target_info = self._set_targets(y)
```
I think you're right here.  Now `get_features_names_and_dtypes` is called from `fit` and passed to `_validate_and_promote_targets`.  Good catch.

grovduck commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
+            # 2. If the target is a string dtype, but has non-string elements,
+            #    this represents mixed types that shouldn't be allowed.
+            if str(dtype) == "category":
+                target = np.asarray(target.tolist())
```
This one is another super tricky one.  Assume this:
```
X = np.array([[1, 2], [3, 4], [5, 6]])
y = pd.Series([True, "bob", 3.0], dtype="category")
RFNodeTransformer().fit(X, y)
```
The dtype associated with the target is `pd.CategoricalDtype` because we want to honor those calls from the user.  Consequently, the promoted dtype (in this case `<U32`) isn't captured as a result from `get_feature_dtypes`.  If we `pass` here, target remains as a `np.object` array and `type_of_target` will return "unknown".  Running the above chunk should error in that way is `pass` is in place.  If we instead promote to "<32", `type_of_target` returns "multiclass".

But thiis gets back to having mixed dtypes in targets.  Because this is a pd.Series and is _explicitly_ labeled as a category, we are currently letting this pass through, whereas if `y` was a numpy array with the same data (cast as np.object), it would be caught as an error.  I'm totally willing to change my mind on this.  How do you feel?

aazuspan replied:
Very tricky, but your explanation makes sense. I knew I was probably missing an edge case that you'd already thought of ;)

> Because this is a pd.Series and is explicitly labeled as a category, we are currently letting this pass through, whereas if y was a numpy array with the same data (cast as np.object), it would be caught as an error.

I totally agree with that logic.

aazuspan commented on src/sknnr/transformers/_rfnode_transformer.py:
```diff
@@ -126,37 +153,125 @@ def __init__(
+        After getting target names and types, validate each target in `y`,
+        ensuring that no targets have NaN-like values and that no targets
+        have a mix of numeric and non-numeric types.  Return the targets as a
+        like of numpy arrays as well as the target names and dtypes.
```
Looks great, thanks!

grovduck commented:
@aazuspan, circling back on the test of non-default parameters (which I think is the last thing to consider in this morass of a PR!).  I've put your test in there (more or less verbatim) for ensuring that the different RF forest types correctly receive specialized (and non-default) parameters.  Since I generally suck at tests, I wasn't sure how far to go with the combinations, but currently I'm testing at least one non-default parameter for each of the five specialized keywords.  Also, I'm not testing combinations of default and non-default parameters which might be valuable?

> I'm not sure we'd gain anything by parametrizing over criterion_reg in the regression test. As long as we're correctly passing it to the forest, I think it's up to sklearn to use it.

If you think we need to do additional regression tests to capture different parameter combinations, I'm happy to do that as well, but I gained from this last comment that you didn't think that was necessary?

aazuspan reviewed (commented):
The parameterization test looks good, and I don't see a strong reason to add any more combinations. The only remaining thing that I see is that the `RFNNRegressor` and `RFNodeTransformer` docstrings don't describe that they infer forest types. That description would presumably need to change once you add in the forest type overrides, so if you'd rather wait for that to finalize the docstrings, that's fine with me.

This was a big feature with a ton of tricky edge cases and sharp corners, so well done sorting through all that and coming up with a great solution!

grovduck commented:
> The only remaining thing that I see is that the `RFNNRegressor` and `RFNodeTransformer` docstrings don't describe that they infer forest types. That description would presumably need to change once you add in the forest type overrides, so if you'd rather wait for that to finalize the docstrings, that's fine with me.

Ah, great catch.  I totally overlooked this one.  Yes, you're right that this will need to change once we have user-overrides in there, but I'd like to update them now to be consistent with this PR.

Thank you for your incredibly patient review on this one.  If you think it's ready, can you accept changes in your review and I'll merge it.

aazuspan reviewed (approved):
LGTM!
